internal/provider/common/errors_test.go

```diff
@@ -67,53 +70,106 @@ func (s *ErrorsSuite) TestNoValidation(c *gc.C) {
 	isAuthF := func(e error) bool {
 		return true
 	}
-	denied := common.MaybeHandleCredentialError(isAuthF, authFailureError, envcontext.WithoutCredentialInvalidator(context.Background()))
-	c.Assert(c.GetTestLog(), jc.DeepEquals, "")
-	c.Assert(denied, jc.IsTrue)
+	denied := common.HandleCredentialError(context.Background(), nil, isAuthF, authFailureError)
+	c.Assert(c.GetTestLog(), jc.Contains, "no credential invalidator provided")
+	c.Check(denied, jc.IsFalse)
 }
 
 func (s *ErrorsSuite) TestInvalidationCallbackErrorOnlyLogs(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
 	isAuthF := func(e error) bool {
 		return true
 	}
-	ctx := envcontext.WithCredentialInvalidator(context.Background(), func(_ context.Context, msg string) error {
-		return errors.New("kaboom")
-	})
-	denied := common.MaybeHandleCredentialError(isAuthF, authFailureError, ctx)
+
+	s.credentialInvalidator.EXPECT().InvalidateCredentials(gomock.Any(), gomock.Any()).Return(errors.New("boom"))
+
+	denied := common.HandleCredentialError(context.Background(), s.credentialInvalidator, isAuthF, authFailureError)
 	c.Assert(c.GetTestLog(), jc.Contains, "could not invalidate stored cloud credential on the controller")
-	c.Assert(denied, jc.IsTrue)
+	c.Check(denied, jc.IsTrue)
 }
 
 func (s *ErrorsSuite) TestHandleCredentialErrorPermissionError(c *gc.C) {
-	s.checkPermissionHandling(c, authFailureError, true)
+	defer s.setupMocks(c).Finish()
 
-	e := errors.Trace(authFailureError)
-	s.checkPermissionHandling(c, e, true)
+	isAuthF := func(e error) bool {
+		return errors.Is(e, authFailureError)
+	}
 
-	e = errors.Annotatef(authFailureError, "more and more")
-	s.checkPermissionHandling(c, e, true)
-}
+	var called bool
+	s.credentialInvalidator.EXPECT().InvalidateCredentials(gomock.Any(), gomock.Any()).DoAndReturn(func(ctx context.Context, reason environs.CredentialInvalidReason) error {
+		c.Assert(string(reason), gc.Matches, "cloud denied access:.*auth failure")
+		called = true
```

Comment:
I don't think `called` is required here? gomock fails if you have an EXPECT that isn't calls the expected number of times.

Reply:
Repeated below

Reply:
It's the inverse, did it call the method? I know it should catch it, but this 100% checks it. Having an `Assert`, `Check` in a mock `Do` and `DoAndReturn` is a bit of smell! We should be setting the reason to a variable out of the scope and then check that at the bottom of the method.